File: tests/test_rtk_cross_client.py
#!/usr/bin/env python3
"""Cross-client RTK verification test across all 5 clients:
Antigravity, Claude Code, Cursor, OpenCode, and CommandCode.
"""
import os
import subprocess
import unittest
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
HOME = Path.home()

class TestRtkCrossClient(unittest.TestCase):
    def test_rtk_binary_detection_and_path(self):
        """Verify rtk binary is detected via Node platform_utils."""
        cmd = [
            "node",
            "-e",
            """
            const { getRtkCommand, isRtkInstalled } = require('./src/platform_utils');
            const rtkCmd = getRtkCommand();
            const installed = isRtkInstalled();
            if (!installed || !rtkCmd) {
                console.error('RTK not detected');
                process.exit(1);
            }
            console.log('RTK detected:', rtkCmd);
            """
        ]
        res = subprocess.run(cmd, cwd=ROOT, capture_output=True, text=True)
        self.assertEqual(res.returncode, 0, f"RTK detection failed: {res.stderr}")
        self.assertIn("RTK detected:", res.stdout)

    def test_rtk_deployment_antigravity(self):
        """Verify RTK deployment to Antigravity CLI and IDE."""
        cmd = [
            "node",
            "-e",
            """
            const { deployAntigravityRtkRule } = require('./src/antigravity_manager');
            const res = deployAntigravityRtkRule(true);
            if (!res.ok) {
                console.error('Antigravity RTK deployment failed:', res);
                process.exit(1);
            }
            """
        ]
        res = subprocess.run(cmd, cwd=ROOT, capture_output=True, text=True)
        self.assertEqual(res.returncode, 0, res.stderr)
        
        # Check rule files exist
        self.assertTrue((HOME / ".gemini" / "antigravity-cli" / "rules" / "rtk.md").exists())
        self.assertTrue((HOME / ".gemini" / "antigravity-ide" / "rules" / "rtk.md").exists())

    def test_rtk_deployment_claude_code(self):
        """Verify RTK deployment and hook registration for Claude Code."""
        cmd = [
            "node",
            "-e",
            """
            const { deployClaudeCodeRtkRule, initRtkHook } = require('./src/mcp_clients_manager');
            const r1 = deployClaudeCodeRtkRule(true);
            const r2 = initRtkHook(true);
            if (!r1.ok || !r2.ok) {
                console.error('Claude Code RTK deployment failed:', r1, r2);
                process.exit(1);
            }
            """
        ]
        res = subprocess.run(cmd, cwd=ROOT, capture_output=True, text=True)
        self.assertEqual(res.returncode, 0, res.stderr)
        self.assertTrue((HOME / ".claude" / "rules" / "rtk.md").exists())

    def test_rtk_deployment_cursor(self):
        """Verify RTK deployment for Cursor."""
        cmd = [
            "node",
            "-e",
            """
            const { deployCursorRtkRule } = require('./src/cursor_manager');
            const res = deployCursorRtkRule(true);
            if (!res.ok) {
                console.error('Cursor RTK deployment failed:', res);
                process.exit(1);
            }
            """
        ]
        res = subprocess.run(cmd, cwd=ROOT, capture_output=True, text=True)
        self.assertEqual(res.returncode, 0, res.stderr)
        self.assertTrue((HOME / ".cursor" / "rules" / "rtk.mdc").exists())

    def test_rtk_deployment_opencode(self):
        """Verify RTK deployment for OpenCode."""
        cmd = [
            "node",
            "-e",
            """
            const { deployOpenCodeRtkRule } = require('./src/opencode_manager');
            const res = deployOpenCodeRtkRule(true);
            if (!res.ok) {
                console.error('OpenCode RTK deployment failed:', res);
                process.exit(1);
            }
            """
        ]
        res = subprocess.run(cmd, cwd=ROOT, capture_output=True, text=True)
        self.assertEqual(res.returncode, 0, res.stderr)
        opencode_rule = (HOME / ".config" / "opencode" / "rules" / "rtk.md")
        if not opencode_rule.exists():
            opencode_rule = (HOME / ".opencode" / "rules" / "rtk.md")
        self.assertTrue(opencode_rule.exists())

    def test_rtk_deployment_commandcode(self):
        """Verify RTK deployment and permissions for CommandCode."""
        cmd = [
            "node",
            "-e",
            """
            const { deployCommandCodeRtkRule, registerCommandCodePermissions } = require('./src/mcp_clients_manager');
            const r1 = deployCommandCodeRtkRule(true);
            registerCommandCodePermissions(true);
            if (!r1.ok) {
                console.error('CommandCode RTK deployment failed:', r1);
                process.exit(1);
            }
            """
        ]
        res = subprocess.run(cmd, cwd=ROOT, capture_output=True, text=True)
        self.assertEqual(res.returncode, 0, res.stderr)
        self.assertTrue((HOME / ".commandcode" / "rules" / "rtk.md").exists())

    def test_rtk_contract_in_rules_and_prompts(self):
        """Verify contract rules require RTK prefixing across rule templates."""
        gemini_rule = (ROOT / "GEMINI.md").read_text(encoding="utf-8")
        self.assertIn("RTK is mandatory for commands", gemini_rule)
        self.assertIn("prefix shell/command execution with `rtk`", gemini_rule)

if __name__ == "__main__":
    unittest.main()

File: tests/test_rtk_cross_client.js
#!/usr/bin/env node
'use strict';

/**
 * tests/test_rtk_cross_client.js — Cross-client RTK verification test across all 6 clients:
 * Antigravity, Claude Code, Cursor, OpenCode, CommandCode, and Codex.
 */

const assert = require('assert');
const fs = require('fs');
const os = require('os');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');
const HOME = os.homedir();

async function run() {
  console.log('Running test_rtk_cross_client tests...');

  // 1. RTK binary detection
  const { getRtkCommand, isRtkInstalled } = require('../src/platform_utils');
  const rtkCmd = getRtkCommand();
  const installed = isRtkInstalled();
  assert.ok(installed && rtkCmd, 'RTK must be detected');
  console.log('✓ RTK binary detection passed:', rtkCmd);

  // 2. Antigravity RTK deployment
  const { deployAntigravityRtkRule } = require('../src/antigravity_manager');
  const resAgy = deployAntigravityRtkRule(true);
  assert.ok(resAgy && resAgy.ok, 'Antigravity RTK deployment failed');
  assert.ok(fs.existsSync(path.join(HOME, '.gemini', 'antigravity-cli', 'rules', 'rtk.md')));
  assert.ok(fs.existsSync(path.join(HOME, '.gemini', 'antigravity-ide', 'rules', 'rtk.md')));
  console.log('✓ Antigravity RTK deployment passed');

  // 3. Claude Code RTK deployment
  const { deployClaudeCodeRtkRule, initRtkHook } = require('../src/mcp_clients_manager');
  const r1 = deployClaudeCodeRtkRule(true);
  const r2 = initRtkHook(true);
  assert.ok(r1 && r1.ok, 'Claude Code RTK deployment failed');
  assert.ok(r2 && r2.ok, 'Claude Code RTK hook init failed');
  assert.ok(fs.existsSync(path.join(HOME, '.claude', 'rules', 'rtk.md')));
  console.log('✓ Claude Code RTK deployment passed');

  // 4. Cursor RTK deployment
  const { deployCursorRtkRule } = require('../src/cursor_manager');
  const resCursor = deployCursorRtkRule(true);
  assert.ok(resCursor && resCursor.ok, 'Cursor RTK deployment failed');
  assert.ok(fs.existsSync(path.join(HOME, '.cursor', 'rules', 'rtk.mdc')));
  console.log('✓ Cursor RTK deployment passed');

  // 5. OpenCode RTK deployment
  const { deployOpenCodeRtkRule } = require('../src/opencode_manager');
  const resOpenCode = deployOpenCodeRtkRule(true);
  assert.ok(resOpenCode && resOpenCode.ok, 'OpenCode RTK deployment failed');
  const opencodeRule1 = path.join(HOME, '.config', 'opencode', 'rules', 'rtk.md');
  const opencodeRule2 = path.join(HOME, '.opencode', 'rules', 'rtk.md');
  assert.ok(fs.existsSync(opencodeRule1) || fs.existsSync(opencodeRule2));
  console.log('✓ OpenCode RTK deployment passed');

  // 6. CommandCode RTK deployment
  const { deployCommandCodeRtkRule, registerCommandCodePermissions } = require('../src/mcp_clients_manager');
  const resCc = deployCommandCodeRtkRule(true);
  registerCommandCodePermissions(true);
  assert.ok(resCc && resCc.ok, 'CommandCode RTK deployment failed');
  assert.ok(fs.existsSync(path.join(HOME, '.commandcode', 'rules', 'rtk.md')));
  console.log('✓ CommandCode RTK deployment passed');

  // 7. Codex RTK deployment
  const { deployCodexRtkRule } = require('../src/codex_manager');
  const resCodex = deployCodexRtkRule(true);
  assert.ok(resCodex && resCodex.ok, 'Codex RTK deployment failed');
  assert.ok(fs.existsSync(path.join(HOME, '.codex', 'rules', 'rtk.md')));
  console.log('✓ Codex RTK deployment passed');

  // 8. RTK contract in rules and prompts
  const geminiRule = fs.readFileSync(path.join(ROOT, 'GEMINI.md'), 'utf-8');
  assert.ok(geminiRule.includes('RTK is mandatory for commands'));
  assert.ok(geminiRule.includes('prefix shell/command execution with `rtk`'));
  console.log('✓ RTK contract in rules and prompts passed');

  console.log('\nAll test_rtk_cross_client tests passed!');
}

if (require.main === module) {
  run().catch(err => {
    console.error(err);
    process.exit(1);
  });
}

module.exports = { run };
